modules/table/src/main/java/org/apache/ignite/internal/table/distributed/index/IndexMetaStorage.java

```diff
@@ -0,0 +1,367 @@
+        CatalogTableDescriptor catalogTableDescriptor = catalogTableDescriptor(catalogIndexDescriptor.tableId(), catalogVersion);
+
+        if (indexId != catalogTableDescriptor.primaryKeyIndexId()) {
+            updateIndexStatus(parameters);
+        }
+    }
+
+    private void onIndexStopping(StoppingIndexEventParameters parameters) {
+        updateIndexStatus(parameters);
+    }
+
+    private void updateIndexStatus(IndexEventParameters parameters) {
+        CatalogIndexDescriptor catalogIndexDescriptor = catalogIndexDescriptor(parameters.indexId(), parameters.catalogVersion());
+
+        Catalog catalog = catalog(parameters.catalogVersion());
+
+        updateAndSaveToVaultIndexMeta(catalogIndexDescriptor.id(), indexMeta -> {
+            assert indexMeta != null : catalogIndexDescriptor.id();
+
+            return setNewStatus(indexMeta, MetaIndexStatusEnum.convert(catalogIndexDescriptor.status()), catalog);
+        });
+    }
+
+    private void onTableAlter(TableEventParameters parameters) {
+        if (parameters instanceof RenameTableEventParameters) {
+            int catalogVersion = parameters.catalogVersion();
+
+            CatalogTableDescriptor catalogTableDescriptor = catalogTableDescriptor(parameters.tableId(), catalogVersion);
+
+            int indexId = catalogTableDescriptor.primaryKeyIndexId();
+
+            CatalogIndexDescriptor catalogIndexDescriptor = catalogIndexDescriptor(indexId, catalogVersion);
+
+            updateAndSaveToVaultIndexMeta(indexId, indexMeta -> {
+                assert indexMeta != null : indexId;
+
+                return indexMeta.indexName(catalogIndexDescriptor.name());
+            });
+        }
+    }
+
+    private void onLwmChanged(ChangeLowWatermarkEventParameters parameters) {
+        int lwmCatalogVersion = catalogService.activeCatalogVersion(parameters.newLowWatermark().longValue());
+
+        Iterator<IndexMeta> it = indexMetaByIndexId.values().iterator();
+
+        while (it.hasNext()) {
+            IndexMeta indexMeta = it.next();
+
+            if (isShouldToRemove(indexMeta, lwmCatalogVersion)) {
+                it.remove();
+
+                removeFromVault(indexMeta);
+            }
+        }
+    }
+
+    private void recoverIndexMetas() {
+        Catalog catalog = catalog(catalogService.latestCatalogVersion());
+
+        Map<Integer, IndexMeta> fromVault = readAllFromVault();
+        Map<Integer, CatalogIndexDescriptor> fromCatalogLatest = readAllFromCatalog(catalog.version());
+
+        for (CatalogIndexDescriptor catalogIndexDescriptor : fromCatalogLatest.values()) {
+            int indexId = catalogIndexDescriptor.id();
+
+            IndexMeta indexMetaFromVault = fromVault.get(indexId);
+
+            if (indexMetaFromVault == null) {
+                // We did not have time to save at the index creation event.
+                updateAndSaveToVaultIndexMeta(indexId, indexMeta -> new IndexMeta(catalogIndexDescriptor, catalog));
```

Comment:
Yes, I think there is a possibility of losing some change in the directory, maybe I can fix it with the next ticket? 
This PR is quite large, it will be necessary to add more tests for such a situation.

Reply:
Then let's have a TODO